File: server/schemas/analytics.py
"""Analytics schemas — ADDENDUM §R.6."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from schemas.attempt import AttemptSummaryOut


class StudentDashboardOut(BaseModel):
    total_scenarios: int
    completed_scenarios: int
    in_progress_scenarios: int
    avg_score: float
    best_score: float
    total_time_hours: float
    recent_attempts: list[AttemptSummaryOut] = Field(default_factory=list)


class ScoreDistributionOut(BaseModel):
    bins: list[int]
    counts: list[int]


class WeakNodeOut(BaseModel):
    node_id: str
    title: str
    node_type: str
    visit_count: int
    avg_score_pct: float
    most_common_wrong_answer: str | None = None


class PathAnalysisOut(BaseModel):
    correct_path_count: int
    incorrect_path_count: int
    most_common_wrong_node: WeakNodeOut | None = None


class StudentRankingEntry(BaseModel):
    user_id: int
    full_name: str
    score: float
    duration_sec: int
    path: list[str] = Field(default_factory=list)


class TeacherScenarioStatsOut(BaseModel):
    scenario_id: int
    scenario_title: str
    group_id: int | None = None
    group_name: str | None = None
    total_students: int
    completed: int
    in_progress: int
    avg_score: float
    score_distribution: ScoreDistributionOut
    path_analysis: PathAnalysisOut
    weak_nodes: list[WeakNodeOut] = Field(default_factory=list)
    student_ranking: list[StudentRankingEntry] = Field(default_factory=list)


class HeatmapNode(BaseModel):
    id: str
    title: str
    node_type: str
    visit_count: int
    avg_score_pct: float | None = None


class HeatmapEdge(BaseModel):
    source: str
    target: str
    traverse_count: int
    is_correct: bool


class PathHeatmapOut(BaseModel):
    scenario_id: int
    group_id: int | None = None
    total_attempts: int
    nodes: list[HeatmapNode] = Field(default_factory=list)
    edges: list[HeatmapEdge] = Field(default_factory=list)


class AdminStatsOut(BaseModel):
    users_total: int
    students: int
    teachers: int
    admins: int
    scenarios_total: int
    published_scenarios: int
    attempts_today: int
    attempts_total: int
    db_size_mb: float
    last_backup_at: datetime | None = None
    last_backup_age_human: str | None = None

File: client/src/types/analytics.ts
import type { AttemptSummaryOut } from './attempt';

export interface StudentDashboardOut {
  total_scenarios: number;
  completed_scenarios: number;
  in_progress_scenarios: number;
  avg_score: number;
  best_score: number;
  total_time_hours: number;
  recent_attempts: AttemptSummaryOut[];
}

export interface ScoreDistributionOut {
  bins: number[];
  counts: number[];
}

export interface WeakNodeOut {
  node_id: string;
  title: string;
  node_type: string;
  visit_count: number;
  avg_score_pct: number;
  most_common_wrong_answer: string | null;
}

export interface PathAnalysisOut {
  correct_path_count: number;
  incorrect_path_count: number;
  most_common_wrong_node: WeakNodeOut | null;
}

export interface StudentRankingEntry {
  user_id: number;
  full_name: string;
  score: number;
  duration_sec: number;
  path: string[];
}

export interface TeacherScenarioStatsOut {
  scenario_id: number;
  scenario_title: string;
  group_id: number | null;
  group_name: string | null;
  total_students: number;
  completed: number;
  in_progress: number;
  avg_score: number;
  score_distribution: ScoreDistributionOut;
  path_analysis: PathAnalysisOut;
  weak_nodes: WeakNodeOut[];
  student_ranking: StudentRankingEntry[];
}

export interface ActivityDayOut {
  date: string;
  count: number;
}

export interface TeacherActivityOut {
  days: ActivityDayOut[];
  total: number;
}

export interface HeatmapNode {
  id: string;
  title: string;
  node_type: string;
  visit_count: number;
  avg_score_pct: number | null;
}

export interface HeatmapEdge {
  source: string;
  target: string;
  traverse_count: number;
}

export interface PathHeatmapOut {
  scenario_id: number;
  group_id: number | null;
  total_attempts: number;
  nodes: HeatmapNode[];
  edges: HeatmapEdge[];
}

export interface AdminStatsOut {
  users_total: number;
  students: number;
  teachers: number;
  admins: number;
  scenarios_total: number;
  published_scenarios: number;
  attempts_today: number;
  attempts_total: number;
  db_size_mb: number;
  last_backup_at: string | null;
  last_backup_age_human: string | null;
}
